### ISSUE TITLE
action=write for ImageReader

### ISSUE THREAD

**Issue by xgastaldi** [Microsoft/CNTK#143]
Hi,
I am trying to get the class predictions from the ResNet CIFAR-10 examples.
These examples use ImageReader to load the pictures. When I use UCIFastReader to write the output, it gives me a different error rate. I don't have this problem with the BatchNorm example which loads the pictures with UCIFastReader.
Is there a way to write outputs with ImageReader?
I tried the following code but it does not work:

---

output_pred = [
    action = "write"
    modelPath=$ModelDir$/05_ResNet.Eval
    minibatchSize = 128
    NDLNetworkBuilder = [
        networkDescription=$ConfigDir$/05_ResNet.ndl
    ]
    reader=[
        readerType=ImageReader
        file=$DataDir$/test_map.txt
        randomize=None
        outputNodeNames= OutputNodes.z
        features=[
            width=32
            height=32
            channels=3
            cropType=Center
            cropRatio=1
            jitterType=UniRatio
            interpolations=Linear
            meanFile=$ConfigDir$/cifar-10-batches-py/data/CIFAR-10_mean.xml
        ]
        labels=[
            labelDim=10
        ]
    ]
    outputPath="$OutputDir$/pred"

**Comment by eldakms** [comment]
Hi xgastaldi,

ImageReader does not have a write mode, it supports only reads. Could you please provide more information (your config) for the unexpected error rate that you are getting using the UCIFastReader?
Thanks!

**Comment by xgastaldi** [comment]
Ok, here are a couple of tests:

Test1: BatchNorm - Read UCIFastReader - Write UCIFastReader
1) ran python CIFAR_convert.py
2) ran cntk configFile=02_BatchNormConv.cntk
CNTK error prediction after 10 epochs: 28.55% (see 02_BatchNormConv_Train_AddBNEval_Test.log)
3) ran cntk configFile=02_BatchNormConvOutput.cntk
4) opened pred.OutputNodes.z in Excel and compared the predictions with the ground truth
5) Prediction of extracted values is 28,55% (cell N2). This is in line with what CNTK gives me

Test2: ResNet - Read ImageReader - Write UCIFastReader
1) ran python CIFAR_convert.py
2) ran python CifarConverter.py
3) In 03_ResNet.cntk
- Changed DataDir = "$RootDir$" to DataDir = "$RootDir$/cifar-10-batches-py"
In the Train block:
- Changed maxEpochs=160 to maxEpochs=10
- Changed meanFile=$ConfigDir$/CIFAR-10_mean.xml to meanFile=$ConfigDir$/cifar-10-batches-py/data/CIFAR-10_mean.xml
In the Test block:
- Changed randomize=Auto to randomize=None
- Changed meanFile=$ConfigDir$/CIFAR-10_mean.xml to meanFile=$ConfigDir$/cifar-10-batches-py/data/CIFAR-10_mean.xml
4) In 03_ResNet.ndl
changed:
#rn2_1_Wproj = Parameter(cMap2, cMap1, init = fromFile, initFromFilePath = "$Proj16to32Filename$", needGradient = false)
#rn2_1 = ResNetNode2Inc(rn1_3, cMap2, 144, 288, kW, kH, convWScale, convBValue, scValue, expAvg, rn2_1_Wproj)
rn2_1 = ResNetNode2Inc2(rn1_3, cMap1, cMap2, 144, 288, kW, kH, convWScale, 3.5, convBValue, scValue, expAvg)
to:
rn2_1_Wproj = Parameter(cMap2, cMap1, init = fromFile, initFromFilePath = "$Proj16to32Filename$", needGradient = false)
rn2_1 = ResNetNode2Inc(rn1_3, cMap2, 144, 288, kW, kH, convWScale, convBValue, scValue, expAvg, rn2_1_Wproj)
#rn2_1 = ResNetNode2Inc2(rn1_3, cMap1, cMap2, 144, 288, kW, kH, convWScale, 3.5, convBValue, scValue, expAvg)
Same thing for rn3_1 (ResNetNode2Inc2 doesn't work because RowStack does not stack along the depth dimension)
5) ran cntk configFile=03_ResNet.cntk
CNTK error prediction after 10 epochs: 19.07% (see 03_ResNet_Train_AddBNEval_Test.log)
6) ran cntk configFile=03_ResNetOutput.cntk
7) opened pred.OutputNodes.z in Excel and compared the predictions with the ground truth
8) Prediction of extracted values is 43.94% (cell N2). This is not in line with what CNTK gives me.

I thought that the difference could come from the meanfile (UCIFastReader does not seem to to accept meanfiles). To check this, I re-ran the previous example without the mean file this time.

Test3: ResNet without mean file - Read ImageReader - Write UCIFastReader
- CNTK error prediction after 10 epochs: 19.53% (see 03_ResNet_Train_AddBNEval_Test.log)
- Prediction of extracted values is 28.52% (cell N2). This a bit better but still  not in line with what CNTK gives me.

[BatchNorm.zip](https://github.com/Microsoft/CNTK/files/141088/BatchNorm.zip)
[ResNet1.zip](https://github.com/Microsoft/CNTK/files/141090/ResNet1.zip)
[ResNet2.zip](https://github.com/Microsoft/CNTK/files/141089/ResNet2.zip)

**Comment by frankseide** [comment]
I noticed you did not specify to disable randomization in your attached 03_ResNetOutput.cntk. Could that be the reason? (It wouldn’t explain though how you still get quite some accuracy.)

One thing you could try is to write out the labels as well, and use those to compare against. That would eliminate the randomization from the equation.

For writing, there are two modes. If you specify no writer, it will instead write simple text output to outputPath. The output should be 100% equivalent to what you get from a dedicated writer. Since the ImageReader has no writer, you will need this mode, so maybe one thing you could try is whether you can use it with the UCIFastReader example and confirm that both output the same values. If not, we need to look there.

**Comment by xgastaldi** [comment]
Hi Frank,
Thanks for your answer.
I modified 03_ResNetOutput.cntk and added randomization=None. I actually get a wider gap (18% from CNTK, 31.86% from the labels).
I haven't tried to write out labels yet so I'll look at the documentation and will give it a try tomorrow.
Concerning your suggestion, that's actually what I tried to do a couple of days ago but CNTK crashes when I replace the UCIFastReader block by an ImageReader block (see the BatchNorm files attached).
Maybe there is something I am missing in the implementation of the "no writer" mode.

[BatchNorm2.zip](https://github.com/Microsoft/CNTK/files/141701/BatchNorm2.zip)

**Comment by frankseide** [comment]
When you switch to the ImageReader block, please do not specify a “writer” section. Just specify an outputPath. You can then also specify the labels themselves as an output node.

**Comment by xgastaldi** [comment]
Do you see any writer block in the code below? I have an action=write command but no writer =[....] block (this is taken from BatchNorm2.zip).
```

output_pred = [
    action = "write"
    modelPath=$ModelDir$/02_BatchNormConv.Eval
    minibatchSize = 16

    NDLNetworkBuilder = [
        networkDescription=$ConfigDir$/02_BatchNormConv.ndl
    ]
    
    reader=[
        readerType=ImageReader
        file=$DataDir$/cifar-10-batches-py/train_map.txt
        randomize=None
        features=[
            width=32
            height=32
            channels=3
            cropType=Random
            cropRatio=0.8
            jitterType=UniRatio
            interpolations=Linear
        ]
        
        labels=[
            dim=1
            start=0
            labelDim=10
            labelMappingFile=$DataDir$/labelsmap.txt
        ]
    ]

    outputPath="$OutputDir$/pred"
    outputNodeNames=OutputNodes.z
]

```

**Comment by xgastaldi** [comment]
I tried to extract the labels (see code below).
What I get back in pred.labels is an array of 0s of size 79x10 ...

[ResNetLabels.zip](https://github.com/Microsoft/CNTK/files/143059/ResNetLabels.zip)

**Comment by alexeyo26** [comment]
Closed due to lack of feedback. Please, re-open if still valid.

**Comment by pkranen** [comment]
The readers have been entirely reworked since the last post. Closing this one, please try the latest bits and open a new issue if required. Thanks.
